Add a 'Reporting' section, improve 'Making Payments' section

bahmady opened the pull request:
### Context
We need a section on reporting in some form or another on the same hierarchy as the other main sections in the docs 

### Changes proposed in this pull request
* New docs for reporting

* Makes some changes to 'Making payments' section

* Changes directory name for 'Making payments' to 'making_payments' (not 'taking_payments' which was erroneous)

### Guidance to review (__please read__)
I'm going to add some comments myself indicating where I'm unsure of things, so perhaps someone with expert knowledge can kindly help me fill those in. 

Please feel free to be as free with your feedback as you please. 

I'll also need to have this reviewed by a writer before I publish it.

I've put this up in an app "in situ" for testing/review: https://govukpay-reportingdocs.cloudapps.digital/

bahmady commented on source/reporting/index.html.md.erb:
````diff
@@ -0,0 +1,147 @@
+      "method": "POST"
+```
+
+## Refund a payment 
````
Little unsure about this section here as we have en entire h2 section dedicated to this subject matter...

bahmady replied:
From Till: we don't need this section. 

We could also talk about "Get all refunds for a payment" in the API if we don't already.

katiebates commented on source/reporting/index.html.md.erb:
```diff
@@ -0,0 +1,147 @@
+* see information users have entered 
+* search through payments 
+* cancel a payment
+* refund a payment
```
very soon they'll also be able to search for refunds too

bahmady replied:
Should we hold off on this PR until that's finished then?

katiebates replied:
we could just add a new section when it's ready - looks like it'll be easy to add it into this new section

katiebates commented on source/reporting/index.html.md.erb:
````diff
@@ -0,0 +1,147 @@
+    },
+```
+
+Other values of `status` are:  
````
In the API, "failed" "error" "success". i'm not sure if "refunded" is a status or whether it's "refund successful", "refund submitted" etc (as in the admin site). 
BTW - the payment states are different in the admin site. don't know if you want to call that out...

bahmady replied:
I'm starting to see why this section was originally titled "Integration"; I guess the idea being that you could integrate API calls into your service in order to come up with something along the lines of reporting. As to whether we should focus this documentation in that vein or whether we should have an additional section explaining how to do the same things with the admin tool, I will defer to you @katiebates :)

katiebates commented on source/reporting/index.html.md.erb:
```diff
@@ -0,0 +1,147 @@
+* cancel a payment
+* refund a payment
+
+You may want to integrate calls to the GOV.UK Pay API in your service if you
```
yes - at its simplest, reconciliation is just comparing what you think you should have been paid for, and what you were paid. So a service could compare its bank statement with what transactions were taken at -2 days to do reconciliation. it's not even stretching the truth ;)

bahmady replied:
Cool, I'll leave this section as-is then 👍

bahmady commented on source/reporting/index.html.md.erb:
```diff
@@ -0,0 +1,147 @@
+
+Other values of `status` are:  
+
+## See information users have entered
```
This could be clearer that it's about a single payment

bahmady commented on source/reporting/index.html.md.erb:
````diff
@@ -0,0 +1,147 @@
+    },
+```
+
+## Search through payments 
````
Could be called "Generate a list of payments", or use the term "Filter"

bahmady commented on source/reporting/index.html.md.erb:
```diff
@@ -0,0 +1,147 @@
+* page
+* display_size
+
+## Cancel a payment
```
Needs to be in own section OR better yet should be in the 'Making payments' section. 

Could also be better called "Cancel a payment that's in progress".

Can be described in two ways. Users can cancel payments by either:

1. users can cancel a payment by clicking a link on the payment page
2. a service can make an API call to cancel a payment that's in progress

bahmady commented:
@katiebates @tillwirth @heathd @alexbishop1 

I've made numerous changes to this after some feedback. Now viewable in situ (in addition to this PR) at https://govukpay-reportingdocs.cloudapps.digital/

Many thanks—any reviews/further feedback welcome :)

bahmady commented:
Is it worth adding this image to the 'Reporting' section:

https://s3-eu-west-1.amazonaws.com/pay-govuk-documentation/Partner-Architecture-v2.png

alexbishop1 commented on source/making_payments/index.html.md.erb:
```diff
@@ -0,0 +1,124 @@
+section of ``links`` in the JSON body (the same URL is shown in both places).
+This URL contains the GOV.UK Pay ``payment_id`` which uniquely identifies the
+payment. An authenticated GET request to the URL will return information about
+the payment and its status.
```
It feels like this might be a good place to refer the reading to the reporting section

alexbishop1 commented on source/making_payments/index.html.md.erb:
```diff
@@ -0,0 +1,124 @@
+
+You will receive the ``next_url``  to which you should direct the user to
+complete their payment. During the GOV.UK Pay beta, it is only returned in
+response to the initial POST call to create a payment, not on sub. It will
```
No idea what “not on sub” is trying to say (“not on subsequent requests”?). In any case, it’s not true: getting a single payment will give you a new one-time `next_url`.

bahmady replied:
No idea, I've just deleted the line.

alexbishop1 commented on source/making_payments/index.html.md.erb:
```diff
@@ -0,0 +1,124 @@
+progress. Alternatively, a service can make the following API call for a given
+`PaymentId`:
+
+`POST /v1/payments/paymentId/cancel`
```
The placeholder in the URL doesn’t match the one in the prose above

alexbishop1 commented on source/making_payments/index.html.md.erb:
```diff
@@ -0,0 +1,124 @@
+
+`GET /v1/payments/paymentId`
+
+If the JSON response body contains a `"cancel"` field that is not `null`, you
```
This should probably mention that it’s within the `_links` object

alexbishop1 commented on source/reporting/index.html.md.erb:
```diff
@@ -0,0 +1,113 @@
+Before reading this section, you should read the section describing [Making
+payments](/making_payments). 
+
+You can use the GOV.UK Pay API to export data automatically into your finance
```
Pay isn’t really exporting the data into the finance systems (not directly, anyway). It’s more that one can develop software to get the data from the API and import it into a finance system.

alexbishop1 replied:
Not convinced about “automatically” — I think “programmatically” might be a more apt term

alexbishop1 commented on source/reporting/index.html.md.erb:
```diff
@@ -0,0 +1,113 @@
+payments](/making_payments). 
+
+You can use the GOV.UK Pay API to export data automatically into your finance
+systems. This requires you to do some development, but could remove 
```
Unqualified, “development” is a bit vague: I think “software development” or even “software development work” would make it clearer

alexbishop1 commented on source/reporting/index.html.md.erb:
```diff
@@ -0,0 +1,113 @@
+systems. This requires you to do some development, but could remove 
+routine manual operations from your workflow and lead to significant cost
+savings for your organisation. Before you do this, you should have a datastore 
+where you are already recording payments. 
```
Not quite sure what “Before you do this, you should have a datastore where you are already recording payments” is trying to say

bahmady replied:
This was from @heathd: is it safe to just delete it?

alexbishop1 commented on source/reporting/index.html.md.erb:
```diff
@@ -0,0 +1,113 @@
+
+You can use the GOV.UK Pay API to:
+
+* get information about single payments
```
I think “get information about a single payment” matches more precisely with the functionality

alexbishop1 commented on source/reporting/index.html.md.erb:
```diff
@@ -0,0 +1,113 @@
+You can use the GOV.UK Pay API to:
+
+* get information about single payments
+* generate a list of payments 
```
Could add “matching specified criteria” or something to this

alexbishop1 commented on source/reporting/index.html.md.erb:
```diff
@@ -0,0 +1,113 @@
+   href="https://gds-payments.gelato.io/api-explorer/gov-uk-pay-api/versions/1.0.2/general/"
+   target="blank">__Find payment by ID__</a> (link opens in new window). 
+   
+For a valid request, the JSON response body contains a `"state"` field. 
```
Might want to expand on why they might be interested in the state field (e.g. if it’s success they know they can release the service to the user etc.)

alexbishop1 commented on source/reporting/index.html.md.erb:
````diff
@@ -0,0 +1,113 @@
+    },
+```
+
+### See information users have entered for a single payment
````
Not sure if this should be in a separate section to the stuff about status. It’s the same API call but the separation makes it look like two different things.

bahmady replied:
Yeah, realised that when I was writing it: I've changed it 👍

alexbishop1 commented on source/reporting/index.html.md.erb:
```diff
@@ -0,0 +1,113 @@
+   href="https://gds-payments.gelato.io/api-explorer/gov-uk-pay-api/versions/1.0.2/general/"
+   target="blank">__Find payment by ID__</a> (link opens in new window). 
+
+For a valid request, the JSON response body contains the `"card_details"`
```
It might not contain `card_details` if they haven’t got that far

alexbishop1 commented on source/reporting/index.html.md.erb:
```diff
@@ -0,0 +1,113 @@
+   href="https://gds-payments.gelato.io/api-explorer/gov-uk-pay-api/versions/1.0.2/general/"
+   target="blank">__Search payments__</a> (link opens in new window). 
+
+You can search payments using the following criteria:
```
We may want to mention that all the criteria for a payment must match to be included in the returned list

bahmady replied:
Ah OK, I didn't know that. So users have to input _all_ criteria to get a return?

alexbishop1 replied:
I meant that if your search criteria specifies, say, the reference and the last four digits of the card number, a payment will only be returned if both the reference _and_ the last four digits match (it’s an “and” search, not an “or” search).

bahmady replied:
Gotcha, will amend, thanks.

alexbishop1 commented on source/reporting/index.html.md.erb:
```diff
@@ -0,0 +1,113 @@
+* `from_date`
+* `to_date`
+* `page`
+* `display_size`
```
There are more than these so maybe the phrasing that introduces the list should state that they are examples of what can be searched for.

The `page` and `display_size` aren’t really criteria so probably need to explained separately (`display_size` is how many results to show at once, `page` is the page of results you want — so if `display_size` is `500` and `page` is `2` and the criteria match 1,543 payments, matching payments 501 through 1000 will be returned).

Might also be worth explaining `from_date` and `to_date` separately (technically a filter but not usually used in the same way as the others). We should state that these dates relate to when the payments were created too.

bahmady replied:
These are the criteria I copied from the API explorer. Do we have an exhaustive list documented anywhere or is there a way to probe the API to find what can be searched? If so I wonder if it would be better to omit examples altogether for now.

alexbishop1 replied:
Currently we have `"reference"`, `"email"`, `"state"`, `"agreement_id"` (only applicable for Direct Debit, so we might want to exclude this one), `"card_brand"`, `"first_digits_card_number"`, `"last_digits_card_number"`, `"cardholder_name"`, `"from_date"`, and `"to_date"`

bahmady commented:
Thanks @alexbishop1: I've gone through and implemented much of your feedback. I still have some questions (where relevant I've added them to your comments). 

The "Making payments" section is largely a holdover and written by a former writer/software person, and it seems like much of it is out of date.

bahmady commented:
@alexbishop1 I've added a note attempting to describe pagination and filtering by date. I can't help but wonder if at least some of this information would be best in the API reference section a la https://docs.registers.service.gov.uk/api_reference/ but I'm happy to revisit at a later time: just want to get some valid content about reporting out. 

Just want to sanity check the API calls noted in the content are correct, too.

alexbishop1 commented on source/reporting/index.html.md.erb:
```diff
@@ -0,0 +1,120 @@
+Before reading this section, you should read the section describing [Making
+payments](/making_payments). 
+
+You can develop software to get data from the GOV.UK Pay API and import it into 
```
Partially contradicting myself earlier, I think this now has too much of an emphasis on software development, which isn’t strictly necessary to use the API for one-off things.

Perhaps it could say something like this (which also makes the example more concrete): “You can get data from the GOV.UK Pay API in a format that’s suitable for automatic processing. You could, for example, develop software to periodically query the GOV.UK Pay API for payment data and import it into your finance system. This could remove routine manual operations from your workflow and lead to significant cost savings for your organisation”

bahmady replied:
> Perhaps it could say something like this (which also makes the example more concrete): “You can get data from the GOV.UK Pay API in a format that’s suitable for automatic processing. You could, for example, develop software to periodically query the GOV.UK Pay API for payment data and import it into your finance system. This could remove routine manual operations from your workflow and lead to significant cost savings for your organisation”

Makes sense, I'll go with that as a base for a new version.

bahmady replied:
Just to be clear about what I meant by that, I'll basically use that content you've just written but restyle it a bit to fit it with GDS copy/style guidance.

alexbishop1 commented on source/reporting/index.html.md.erb:
```diff
@@ -0,0 +1,120 @@
+
+## Get information about a single payment
+
+`GET /v1/payments/paymentId`
```
This is fine — not sure if we have a convention for indicating that `paymentId` is a placeholder

bahmady replied:
We should start one: I'd like to look at that separately across the whole document soon after we eventually merge this PR.

bravegrape replied:
@bahmady I'd also be interested in what you decide for this 👍

alexbishop1 commented on source/reporting/index.html.md.erb:
```diff
@@ -0,0 +1,120 @@
+   href="https://gds-payments.gelato.io/api-explorer/gov-uk-pay-api/versions/1.0.2/general/"
+   target="blank">__Find payment by ID__</a> (link opens in new window). 
+   
+For a valid request, the JSON response body contains a `"state"` field. You
```
Thinking about it, “If the payment exists” is probably more specific than “For a valid request”

bahmady replied:
Sounds sensible.

alexbishop1 commented on source/reporting/index.html.md.erb:
```diff
@@ -0,0 +1,120 @@
+
+## Generate a list of payments (search payments) 
+
+`GET /v1/payments`
```
This is fine (it will get all payments) — maybe we should explicitly call out that all payments are returned if no query parameters are provided

bahmady replied:
> maybe we should explicitly call out that all payments are returned if no query parameters are provided

Sounds like a good idea.

alexbishop1 commented on source/reporting/index.html.md.erb:
```diff
@@ -0,0 +1,120 @@
+* `card_brand` (example: `master-card`)
+* `first_digits_card_number`
+* `last_digits_card_number`
+* `cardholder_name`
```
It might be useful to have an example after this (something like `GET /v1/payments?from_date=2018-10-01T13:30:00Z&to_date=2018-10-14T15:00:00Z&reference=12345&state=success`)

alexbishop1 commented on source/reporting/index.html.md.erb:
```diff
@@ -0,0 +1,120 @@
+### Pagination
+
+You can use the `page` and `display-size` parameters for pagination. These
+default to 1 and must be a positive integer. For example, for a search
```
`display-size` defaults to 500

bahmady replied:
Whoops.

alexbishop1 commented on source/reporting/index.html.md.erb:
```diff
@@ -0,0 +1,120 @@
+matching 1543 payments where `display-size` is 500 and `page` is `2`, the
+result will return payments 501-1000. 
+
+### Filter by date
```
Filtering by date is very common and I think we want to encourage people to use it to avoid getting tonnes of payments, so I we might want this bit to be before the other criteria parameters bit

bahmady replied:
Sounds sensible.

alexbishop1 commented on source/reporting/index.html.md.erb:
```diff
@@ -0,0 +1,120 @@
+`to_date` is the end date for payments to be searched, exclusive. 
+
+If you search for a specific payment, all criteria you use must match.
+Otherwise, that payment will not be returned in the results. 
```
Maybe this should go with the criteria parameters bit? It could be taken that it only applies to when date filters are used basked on its position in the document.

bahmady replied:
Makes sense.

bahmady commented:
Thanks @alexbishop1, I've added all your suggestions per my comments against each 👍 

Let me know what you think of the new version. The only thing I haven't added is something to call out placeholders explicitly: we do need that, but I'd like to make it uniform across the whole docs, so I'd like to do that separately in a different PR ideally.

bahmady commented on source/reporting/index.html.md.erb:
```diff
@@ -0,0 +1,135 @@
+weight: 70
+---
+
+# Reporting 
```
"Reporting" is a term in finance, collocated with "Reconciliation", so it's not strictly about generating a report or reports so much as it is about how you can use the admin tool or API to do something which users should recognise from the title.

bahmady commented on source/reporting/index.html.md.erb:
````diff
@@ -0,0 +1,135 @@
+    },
+```
+
+## Generate a list of payments (search payments) 
````
Yeah—it's a slightly tricky one but as I'll address in some other replies, we're trying to meet the user need of being recognisable to non-technical product or service managers whilst also providing instructions for technical users responsible for integrations with the API.

bahmady commented on source/reporting/index.html.md.erb:
```diff
@@ -0,0 +1,135 @@
+
+# Reporting 
+
+You can extract reporting information from GOV.UK Pay manually, or via the
```
I'll duplicate a comment from elsewhere in answer below but yes, it should be obvious for the intended users:

"Reporting" is a term in finance, collocated with "Reconciliation", so it's not strictly about generating a report or reports so much as it is about how you can use the admin tool or API to do something which users should recognise from the title.

bahmady commented on source/reporting/index.html.md.erb:
```diff
@@ -0,0 +1,135 @@
+
+## Reporting via the GOV.UK Pay API 
+
+Before reading this section, you should read the section describing [Making
```
I like this pattern, thanks 👍

bahmady commented on source/reporting/index.html.md.erb:
```diff
@@ -0,0 +1,135 @@
+You can filter transactions using the available search criteria, and then
+select __Download all transaction details as a CSV file__. 
+
+You can process CSV files using spreadsheet software.
```
>  Isn't telling them to process CSV using spreadsheet software a bit patronising?

For some of the users, yes, but as I address elsewhere we're trying to meet the user needs of not-necessarily-technical product and service managers (who might not know about what a CSV is vis a vis spreadsheet software). 

> You could turn the last line into something like 'You now have your CSV report which you can process using spreadsheet software'. This is also a good content pattern as it reinforces the outcome and it also has a call to action.

I'm a little bit less sure of this suggestion because I think it might be unnecessarily long and I'm not sure if it needs a call to action in this particular case.

bahmady commented on source/reporting/index.html.md.erb:
```diff
@@ -0,0 +1,135 @@
+account](https://selfservice.payments.service.gov.uk/login) and export
+transaction information in a CSV file. 
+
+For a given account, select __Transactions__ in the __Dashboard__ page.
```
I think "For a given account" isn't a great pattern for this particular context, in hindsight. It would be obvious to the user yes, but I think it would be better to say "For the account you want to look at". There may be a better way of saying it like, "For the account you want to extract reporting information for" but I'll let my PM comment on that when she's back if needed. Thanks for picking this up.

bahmady commented on source/reporting/index.html.md.erb:
```diff
@@ -0,0 +1,135 @@
+You can extract reporting information from GOV.UK Pay manually, or via the
+API.
+
+## Manual reporting
```
Yes, agreed on both.

bahmady commented on source/reporting/index.html.md.erb:
```diff
@@ -0,0 +1,135 @@
+
+1. Sign in to <a href="https://gds-payments.gelato.io/api-explorer/"
+   target="blank">the API Explorer</a> (link opens in new window).  
+2. Under __Resource__, select __{Payment Id}. Under __Action__, select <a
```
This one's a little weird (like the others). I toyed around with it but if you actually look at the API Explorer they're right next to each other, which is why they're in a single step. But in hindsight, there's no reason not to have them in separate steps, and it looks nicer.

bahmady commented on source/reporting/index.html.md.erb:
```diff
@@ -0,0 +1,135 @@
+   href="https://gds-payments.gelato.io/api-explorer/gov-uk-pay-api/versions/1.0.2/general/"
+   target="blank">__Find payment by ID__</a> (link opens in new window). 
+   
+If a payment exists, the JSON response body to this API call contains a
```
Agree re. "if _the_ payment exists"

bahmady commented on source/reporting/index.html.md.erb:
```diff
@@ -0,0 +1,135 @@
+2. Under __Resource__, select __{Payment Id}. Under __Action__, select <a
+   href="https://gds-payments.gelato.io/api-explorer/gov-uk-pay-api/versions/1.0.2/general/"
+   target="blank">__Find payment by ID__</a> (link opens in new window). 
+   
```
Normally I'd agree but it's a little hard with this particular API explorer format: I can show you what I mean when we meet tomorrow in person.

bahmady commented on source/reporting/index.html.md.erb:
```diff
@@ -0,0 +1,135 @@
+   target="blank">__Find payment by ID__</a> (link opens in new window). 
+   
+If a payment exists, the JSON response body to this API call contains a
+`"state"` field. You can use this information when designing a service.  
```
> I'd recommend having separate section dedicated to useful information in a payment report. 

That's likely a good idea; I'll talk to my PM about this when she's back.

bahmady commented on source/reporting/index.html.md.erb:
```diff
@@ -0,0 +1,135 @@
+If you search for a specific payment, all criteria you use must match.
+Otherwise, that payment will not be returned in the results. 
+
+An example search request:
```
Based on the users for this particular section, it should be obvious what each aspect of the search example means in relation to query parameters, but I'll definitely keep an eye on user research and if people do need explicit explanations.

bahmady commented:
I'm trying to be a bit creative with the ordering of the search/query parameters: does it look better/worse like this?

katiebates commented:
Looks good to me. A few comments: 

1. Depending on how 'reporting' is understood, we could expand this section. 
A - the dashboard is a form of reporting. service managers tend to use this to look at aggregate figures. however, it's pretty simple/ obvious at the moment so maybe it doesn't need explanation. the only thing which is perhaps not 100% obvious is the date ranges - for example, 'last 30 days' doesn't include today's transactions. (see https://govuk.zendesk.com/agent/tickets/2789063). 
B - i wonder if the section makes it sound like you can only download a CSV, and not also see a list of transactions? just having a list of transactions is useful for reporting too. feel like i'm splitting hairs here though 

2. CSV - although the title for the download button is "Download all transaction details as a CSV file." - we might want to spell out that the CSV gives all transaction data (we find that some services then edit it down to remove data they don't need), so they might want to be a bit circumspect about routinely downloading and keeping the files since there's a fair bit of PII in them. 

3. page size - the maximum is 500 - it's not just the default. The API returns an error if a request is made with a display_size parameter > 500.

4. Is the display page example correct - why would page 1 display transactions 501-1000? Wouldn't it be 1-500?
